### ISSUE TITLE
qBittorrent not gracefully closing when container is stopped

### ISSUE THREAD

**Issue by thrnz** [qbittorrent/docker-qbittorrent-nox#9]
When stopped, the container currently doesn't appear to wait for qBittorrent to close gracefully before exiting. On shutdown, once `entrypoint.sh` closes the container exits and the qbt process is instantly killed.

Using `exec` in `entrypoint.sh` when launching qbt appears to be enough to fix this:

```
exec doas -u qbtUser \
    qbittorrent-nox \
        --profile="$profilePath" \
        --webui-port="$QBT_WEBUI_PORT" \
        "$@"
```

The `entrypoint.sh` process then gets replaced by `qbittorrent-nox`. I guess `tini` might only be waiting for the main child process to finish before exiting.

**Comment by Chocobo1** [comment]
>When stopped, the container currently doesn't appear to wait for qBittorrent to close gracefully before exiting. 

Is there a way to verify this?

**Comment by Chocobo1** [comment]
>On shutdown, once entrypoint.sh closes the container exits and the qbt process is instantly killed.

I could be missing something here. Isn't `entrypoint.sh` waiting for qbt? otherwise qbt won't live long enough for users to use. 
And once qbt is exited, `entrypoint.sh` ends too so there is no problem?

**Comment by thrnz** [comment]
> > When stopped, the container currently doesn't appear to wait for qBittorrent to close gracefully before exiting.
> 
> Is there a way to verify this?

It shows up in the `qbittorrent.log` file. Here's a normal, clean shutdown:

```
(N) 2023-05-24T05:29:56 - qBittorrent termination initiated
(I) 2023-05-24T05:29:56 - UPnP/NAT-PMP support: OFF
(N) 2023-05-24T05:30:37 - qBittorrent is now ready to exit
```
And here's what currently happens when the container is stopped. qbt seems to receive the `SIGTERM`, but is killed shortly after:

```
(N) 2023-05-24T05:34:11 - qBittorrent termination initiated
```

> FYI, it is strongly recommended to set `--stop-timeout 1800` when you use `docker run`.

I'm using the equivalent docker-compose setting: `stop_grace_period: 600s`, so timeouts hopefully aren't an issue here.

> > On shutdown, once entrypoint.sh closes the container exits and the qbt process is instantly killed.
> 
> I could be missing something here. Isn't `entrypoint.sh` waiting for qbt? otherwise qbt won't live long enough for users to use. And once qbt is exited, `entrypoint.sh` ends too so there is no problem?

It is waiting for qbt, and it works fine when qbt is closed from the web interface - `entrypoint.sh` simply ends once qbt has finished shutting down and the container exits. When the container is brought down from the command line however (eg. `docker-compose down`), `entrypoint.sh` seems to exit before qbt and it seems that `tini` doesn't wait for qbt to also exit before the container goes down and qbt just gets prematurely killed.

I first noticed things might be wonky when the container was closing down instantly. Note the time delay in the log file during the successful graceful shutdown while fastresume stuff is written.

**Comment by Chocobo1** [comment]
> It shows up in the `qbittorrent.log` file. Here's a normal, clean shutdown:
> 
> ```
> (N) 2023-05-24T05:29:56 - qBittorrent termination initiated
> (I) 2023-05-24T05:29:56 - UPnP/NAT-PMP support: OFF
> (N) 2023-05-24T05:30:37 - qBittorrent is now ready to exit
> ```
> 
> And here's what currently happens when the container is stopped. qbt seems to receive the `SIGTERM`, but is killed shortly after:
> 
> ```
> (N) 2023-05-24T05:34:11 - qBittorrent termination initiated
> ```

I confirm this. Will look into it.

**Comment by Chocobo1** [comment]
@thrnz 
I've landed your suggested fix in a27e526829a196c1aa852f228d4d157214fde2dd.
I plan to release a new image later if there are no other issues.

Thank you for the help!
